grammar/haxe.bnf

```diff
@@ -325,8 +325,9 @@ classDeclaration ::= macroClassList? privateKeyWord? 'class' componentName gener
 
 //'from' | 'to'
 underlyingType ::= '(' typeWrapper ')'
-abstractClassDeclaration ::= macroClassList? privateKeyWord? 'abstract' componentName genericParam? underlyingType? ((identifier) type)* '{' abstractBody '}'
+abstractClassDeclaration ::= macroClassList? privateKeyWord? abstractClassType componentName genericParam? underlyingType? ((identifier) type)* '{' abstractBody '}'
 {pin=3 mixin="com.intellij.plugins.haxe.lang.psi.impl.AbstractHaxePsiClass" implements="com.intellij.plugins.haxe.lang.psi.HaxeClass"}
+abstractClassType ::= ('enum' 'abstract' | 'abstract')
```

Comment:
Why not make this private?

Also, don't you want the rule to be?:
private abstractClassType ::= ('enum' 'abstract') | 'abstract'

or possibly?:
private abstractClassType ::= 'enum'? 'abstract'

Reply:
The rule syntax you have works fine, though the last one is easier to read.  I was thinking that the '|' bound only to the nearest token, rather than entire expressions.

Reply:
> Why not make this private?
Reason is simple - it's hard to extract right token from ast-tree if there is no direct access for it.